src/BuiltInTools/dotnet-watch/HotReload/NamedPipeContract.cs

```diff
@@ -79,117 +63,369 @@ public static void WriteLog(Stream stream, IReadOnlyCollection<(string message,
+        }
+
+        /// <summary>
+        /// Called by delta applier.
+        /// </summary>
+        public static async IAsyncEnumerable<(string message, AgentMessageSeverity severity)> ReadLogAsync(Stream stream, [EnumeratorCancellation] CancellationToken cancellationToken)
+        {
+            var entryCount = await stream.ReadInt32Async(cancellationToken);
 
-            static async ValueTask<byte[]> ReadBytesAsync(BinaryReader binaryReader, CancellationToken cancellationToken)
+            for (var i = 0; i < entryCount; i++)
             {
-                var numBytes = binaryReader.ReadInt32();
+                var message = await stream.ReadStringAsync(cancellationToken);
+                var severity = (AgentMessageSeverity)await stream.ReadByteAsync(cancellationToken);
+                yield return (message, severity);
+            }
+        }
+    }
 
-                var bytes = new byte[numBytes];
+    internal readonly struct ClientInitializationPayload(string capabilities)
+    {
+        private const byte Version = 0;
 
-                var read = 0;
-                while (read < numBytes)
-                {
-                    read += await binaryReader.BaseStream.ReadAsync(bytes.AsMemory(read), cancellationToken);
-                }
+        public string Capabilities { get; } = capabilities;
+
+        /// <summary>
+        /// Called by delta applier.
+        /// </summary>
+        public async ValueTask WriteAsync(Stream stream, CancellationToken cancellationToken)
+        {
+            await stream.WriteAsync(Version, cancellationToken);
+            await stream.WriteAsync(Capabilities, cancellationToken);
+        }
+
+        /// <summary>
+        /// Called by dotnet-watch.
+        /// </summary>
+        public static async ValueTask<ClientInitializationPayload> ReadAsync(Stream stream, CancellationToken cancellationToken)
+        {
+            var version = await stream.ReadByteAsync(cancellationToken);
+            if (version != Version)
+            {
+                throw new NotSupportedException($"Unsupported version {version}.");
+            }
+
+            var capabilities = await stream.ReadStringAsync(cancellationToken);
+            return new ClientInitializationPayload(capabilities);
+        }
+    }
+
+    internal readonly struct StaticAssetPayload(
+        string assemblyName,
+        string relativePath,
+        byte[] contents,
+        bool isApplicationProject)
+    {
+        private const byte Version = 1;
+
+        public string AssemblyName { get; } = assemblyName;
+        public bool IsApplicationProject { get; } = isApplicationProject;
+        public string RelativePath { get; } = relativePath;
+        public byte[] Contents { get; } = contents;
+
+        public async ValueTask WriteAsync(Stream stream, CancellationToken cancellationToken)
+        {
+            await stream.WriteAsync(Version, cancellationToken);
```

Comment:
Are there comments somewhere that describe how the bytes are supposed to be laid out?

Reply:
It can be anything as long as we have the same code on both sides.